packages/audiofileplayer/android/src/main/java/com/google/flutter/plugins/audiofileplayer/AudiofileplayerPlugin.java

```diff
@@ -112,37 +121,103 @@
   // Used when defining an Intent from a custom media button.
   public static final String CUSTOM_MEDIA_BUTTON_EXTRA_KEY = "customMediaButton";
 
-  private final Registrar registrar;
-  private final Map<String, ManagedMediaPlayer> mediaPlayers;
-  private final MethodChannel methodChannel;
+  private Registrar registrar;
+  private Activity activity;
+  private Map<String, ManagedMediaPlayer> mediaPlayers;
+  private MethodChannel methodChannel;
+  private Context context;
+  private ActivityPluginBinding activityPluginBinding;
 
   private MediaBrowserCompat mediaBrowser;
   private MediaControllerCompat mediaController;
 
   public static void registerWith(Registrar registrar) {
-    final MethodChannel methodChannel = new MethodChannel(registrar.messenger(), CHANNEL);
-    final AudiofileplayerPlugin instance = new AudiofileplayerPlugin(registrar, methodChannel);
-    methodChannel.setMethodCallHandler(instance);
+    AudiofileplayerPlugin instance = new AudiofileplayerPlugin();
+    instance.registrar = registrar;
+    instance.activity = registrar.activity();
+    instance.initInstance(registrar.messenger(), registrar.context());
+    instance.registerLifecycleCallbacks(instance.activity);
   }
 
-  private AudiofileplayerPlugin(Registrar registrar, MethodChannel methodChannel) {
-    this.registrar = registrar;
-    this.methodChannel = methodChannel;
-    this.mediaPlayers = new HashMap<>();
-    LifecycleCallbacks callbacks = new LifecycleCallbacks(this, registrar.activity().hashCode());
-    registrar.activity().getApplication().registerActivityLifecycleCallbacks(callbacks);
+  private void registerLifecycleCallbacks(Activity activity) {
+    LifecycleCallbacks callbacks = new LifecycleCallbacks(this, activity.hashCode());
+    activity.getApplication().registerActivityLifecycleCallbacks(callbacks);
+  }
 
-    // Set up MediaBrowser to connect to AudiofileplayerService. Service will be started on first
-    // playback of a background asset.
-    Context context = registrar.activeContext();
+  private void initInstance(BinaryMessenger messenger, Context context) {
+    this.context = context;
+    methodChannel = new MethodChannel(messenger, CHANNEL);
+    methodChannel.setMethodCallHandler(this);
+    mediaPlayers = new HashMap<>();
+    Context activeContext = activeContext();
```

Comment:
You can just use context here. At this point, there must not have activity.

Reply:
It only has activity for apps that are still on v1 and register the plugin in foreground. I don't see the point of using activeContext here. But please confirm with @monkeyswarm .

Reply:
What is the different between 'context' and 'activeContext'?

Reply:
we still need activeContext() to switch between application context and activity context for v1. I added the logic to identify case by case for both v1 and v2 within the function.